toolchain/check/deduce.cpp

```diff
@@ -293,36 +306,23 @@ auto DeductionContext::Deduce() -> bool {
   while (!worklist_.Done()) {
     auto [param_id, arg_id, needs_substitution] = worklist_.PopNext();
 
-    auto note_initializing_param = [&](auto& builder) {
-      if (auto param =
-              context().insts().TryGetAs<SemIR::SymbolicBindingPattern>(
-                  param_id)) {
-        CARBON_DIAGNOSTIC(InitializingGenericParam, Note,
-                          "initializing generic parameter `{0}` declared here",
-                          SemIR::NameId);
-        builder.Note(
-            param_id, InitializingGenericParam,
-            context().entity_names().Get(param->entity_name_id).name_id);
-      } else {
-        NoteGenericHere(context(), generic_id_, builder);
-      }
-    };
-
     // TODO: Bail out if there's nothing to deduce: if we're not in a pattern
     // and the parameter doesn't have a symbolic constant value.
 
-    // If the parameter has a symbolic type, deduce against that.
     auto param_type_id = context().insts().Get(param_id).type_id();
+    // If the parameter has a symbolic type, deduce against that.
     if (param_type_id.AsConstantId().is_symbolic()) {
       Add(context().types().GetInstId(param_type_id),
           context().types().GetInstId(context().insts().Get(arg_id).type_id()),
           needs_substitution);
     } else {
-      // The argument needs to have the same type as the parameter.
-      // TODO: Suppress diagnostics here if diagnose_ is false.
-      // TODO: Only do this when deducing against a symbolic pattern.
-      DiagnosticAnnotationScope annotate_diagnostics(&context().emitter(),
-                                                     note_initializing_param);
+      // The argument may be a runtime value (e.g. TupleLiteral of types), but
+      // is convertible to a compile-time value (e.g. TupleType). So we do this
+      // conversion here, even though we will later try convert again when we
+      // have deduced all of the bindings.
```

Comment:
I think the "runtime value" / "compile-time value" distinction here is not quite right, because the cases we care about here are primarily the ones where the constant value of the argument is *not* `ConstantId::Runtime` (eg, tuple literals with compile time values). Maybe:
```suggestion
      // The argument (e.g. a TupleLiteral of types) may be convertible to a
      // compile-time value (e.g. TupleType) that we can decompose further.
      // So we do this conversion here, even though we will later try convert
      // again when we have deduced all of the bindings.
```
... if I'm correctly understanding why this is important. (Though if that's right, I guess we'd need to do something comparable when the parameter type is symbolic.)

Just for my information, what goes wrong if we remove this conversion?

Reply:
When the argument is a `TupleLiteral` of types, then `context().constant_values().Get(arg_id)` gives `NotConstant` as the constant value, and thus `context().constant_values().GetConstantInstId()` sees it is not symbolic nor template, and returns `None`. Then `result_arg_ids_[index.index]` is not set at all for the argument. And then a lot of tests fail. If we _do_ the conversion then `TupleLiteral` is converted to `TupleType` and we get a value in `result_arg_ids_[index.index]`.

Here's a concrete example from toolchain/check/testdata/deduce/tuple.carbon. You can see the error we get in the test output:
```
// --- tuple_value.carbon

library "[[@TEST_NAME]]";

class HasPair(Pair:! (i32, i32)) {}

// CHECK:STDERR: tuple_value.carbon:[[@LINE+7]]:27: error: argument for generic parameter is not a compile-time constant [CompTimeArgumentNotConstant]
// CHECK:STDERR: fn F[A:! i32, B:! i32](h: HasPair((A, B))) -> i32 { return B; }
// CHECK:STDERR:                           ^~~~~~~~~~~~~~~
// CHECK:STDERR: tuple_value.carbon:[[@LINE-5]]:15: note: initializing generic parameter `Pair` declared here [InitializingGenericParam]
// CHECK:STDERR: class HasPair(Pair:! (i32, i32)) {}
// CHECK:STDERR:               ^~~~
// CHECK:STDERR:
fn F[A:! i32, B:! i32](h: HasPair((A, B))) -> i32 { return B; }
```
When we get to deducing the `(i32, i32)` parameter of `HasPair` we have these as param and arg:
```
(gdb) call Dump(context(), param_id)
inst49: {kind: SymbolicBindingPattern, arg0: entity_name6, type: type(inst45)}
  - type type(inst45): (i32, i32); {kind: TupleType, arg0: type_block2, type: type(TypeType)}
  - value symbolic_constant22: {inst: inst50, generic: generic1, index: generic_inst_in_decl1, .Self: false}
(gdb) call Dump(context(), arg_id)
inst80: {kind: TupleLiteral, arg0: inst_block26, type: type(inst45)}
  - type type(inst45): (i32, i32); {kind: TupleType, arg0: type_block2, type: type(TypeType)}
  - value runtime
```
You can see that the argument's value is 'runtime' there. The `loc_id_` is `tuple_value.carbon:13:41` which puts us at the call to `HasPair((A, B))` in the type expression of `h:`.

> (Though if that's right, I guess we'd need to do something comparable when the parameter type is symbolic.)

Right... In the example above, the parameter has a symbolic value, but its type is not symbolic. I can try construct an example where it is and see what happens. We should in theory hit this `CompTimeArgumentNotConstant` error.